### PR TITLE
Add Ada Language Server (als)

### PR DESCRIPTION
**Pull request by glacambre**
This pull request adds support for the Ada Language Server (https://github.com/AdaCore/ada_language_server). I mostly copied what happened for the scala language server (metals) so if there are issues with this PR, there might also be some with metals :)

One thing that I'm slightly uncomfortable with is the install function, which I only tested on Linux: if it succeeds on OSX, this would be only by accident and I'm quite convinced that it would fail on Windows. Since other installers (pyls and metals) do not seem to try to be multiplatform, I thought this wouldn't be a huge problem.

### CODE REVIEW

**Comment by glacambre** [comment]
Everything seems to work fine, except for the omnifunc.
With the following file:
```ada
procedure tmp is
        procedure a1 is
        begin
                null;
        end a1;
        procedure a2 is
        begin
                a1;
        end a2;
begin
end tmp;
```
And the following setting:
```vim
setlocal omnifunc=v:lua.vim.lsp.omnifunc 
```
Attempting to write a third procedure `a3` below `a2` results in the following error messages being printed on each keypress:
```
Error executing vim.schedule lua callback: ...t_nvimDWUr16/usr/share/nvim/runtime/lua/vim/lsp/util.lua:701: attempt to concatenate a nil value
```

**Comment by h-michael** [comment]
> Attempting to write a third procedure a3 below a2 results in the following error messages being printed on each keypress:

Would you show me log?

`:lua vim.lsp.set_log_level("debug")`, `:lua print(vim.lsp.get_log_path())`

**Comment by glacambre** [comment]
```
[ DEBUG ] 2020-03-17T06:31:01Z+0100 ] ....mount_nvim7cMn6y/usr/share/nvim/runtime/lua/vim/lsp.lua:608 ]	"on_lines"	1	4	10	11	11	9	9	9	{ "        p" }
[ DEBUG ] 2020-03-17T06:31:01Z+0100 ] ...nt_nvim7cMn6y/usr/share/nvim/runtime/lua/vim/lsp/rpc.lua:269 ]	"rpc.notify"	"textDocument/didChange"	{  contentChanges = { {      text = "\nprocedure tmp is\n        procedure a1 is\n        begin\n                null;\n        end a1;\n        procedure a2 is\n        begin\n                a1;\n        end a2;\n        p\nbegin\n        null;\nend tmp;\n"    } },  textDocument = {    uri = "file:///home/me/prog/ada/tmp.adb",    version = 4  }}
[ DEBUG ] 2020-03-17T06:31:01Z+0100 ] ...nt_nvim7cMn6y/usr/share/nvim/runtime/lua/vim/lsp/rpc.lua:258 ]	"rpc.send.payload"	{  jsonrpc = "2.0",  method = "textDocument/didChange",  params = {    contentChanges = { {        text = "\nprocedure tmp is\n        procedure a1 is\n        begin\n                null;\n        end a1;\n        procedure a2 is\n        begin\n                a1;\n        end a2;\n        p\nbegin\n        null;\nend tmp;\n"      } },    textDocument = {      uri = "file:///home/me/prog/ada/tmp.adb",      version = 4    }  }}
[ DEBUG ] 2020-03-17T06:31:01Z+0100 ] ...nt_nvim7cMn6y/usr/share/nvim/runtime/lua/vim/lsp/rpc.lua:338 ]	"decoded"	{  jsonrpc = "2.0",  method = "textDocument/publishDiagnostics",  params = {    diagnostics = { {        message = "Skipped token p",        range = {          end = {            character = 9,            line = 10          },          start = {            character = 8,            line = 10          }        }      } },    uri = "file:///home/me/prog/ada/tmp.adb"  }}
[ DEBUG ] 2020-03-17T06:31:01Z+0100 ] ....mount_nvim7cMn6y/usr/share/nvim/runtime/lua/vim/lsp.lua:367 ]	"notification"	"textDocument/publishDiagnostics"	{  diagnostics = { {      message = "Skipped token p",      range = {        end = {          character = 9,          line = 10        },        start = {          character = 8,          line = 10        }      }    } },  uri = "file:///home/me/prog/ada/tmp.adb"}
[ DEBUG ] 2020-03-17T06:31:01Z+0100 ] ...m7cMn6y/usr/share/nvim/runtime/lua/vim/lsp/callbacks.lua:252 ]	"default_callback"	"textDocument/publishDiagnostics"	{  client_id = 1,  params = {    diagnostics = { {        message = "Skipped token p",        range = {          end = {            character = 9,            line = 10          },          start = {            character = 8,            line = 10          }        }      } },    uri = "file:///home/me/prog/ada/tmp.adb"  }}
[ INFO ] 2020-03-17T06:31:06Z+0100 ] ....mount_nvim7cMn6y/usr/share/nvim/runtime/lua/vim/lsp.lua:797 ]	"exit_handler"	{ {    _on_attach = <function 1>,    callbacks = <1>{      ["window/logMessage"] = <function 2>,      ["window/showMessage"] = <function 3>,      ["workspace/configuration"] = <function 4>    },    cancel_request = <function 5>,    config = {      callbacks = <table 1>,      capabilities = {        textDocument = {          completion = {            completionItem = {              commitCharactersSupport = false,              deprecatedSupport = false,              documentationFormat = { "markdown", "plaintext" },              preselectSupport = false,              snippetSupport = false            },            completionItemKind = {              valueSet = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25 }            },            contextSupport = false,            dynamicRegistration = false          },          documentHighlight = {            dynamicRegistration = false          },          documentSymbol = {            dynamicRegistration = false,            hierarchicalDocumentSymbolSupport = true,            symbolKind = {              valueSet = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26 }            }          },          hover = {            contentFormat = { "markdown", "plaintext" },            dynamicRegistration = false          },          references = {            dynamicRegistration = false          },          signatureHelp = {            dynamicRegistration = false,            signatureInformation = {              documentationFormat = { "markdown", "plaintext" }            }          },          synchronization = {            didSave = true,            dynamicRegistration = false,            willSave = false,            willSaveWaitUntil = false          }        },        workspace = {          configuration = true,          <metatable> = <2>{            __tostring = <function 6>          }        }      },      cmd = { "/home/me/.cache/nvim/nvim_lsp/als/linux/ada_language_server" },      filetype = { "ada" },      init_options = vim.empty_dict(),      log_level = 2,      message_level = 2,      name = "als",      on_attach = <function 7>,      on_exit = <function 8>,      on_init = <function 9>,      root_dir = "/home/me/prog/ada",      settings = vim.empty_dict(),      <metatable> = <table 2>    },    id = 1,    initialized = true,    is_stopped = <function 10>,    name = "als",    notify = <function 11>,    offset_encoding = "utf-16",    request = <function 12>,    resolved_capabilities = {      code_action = false,      document_formatting = false,      document_highlight = false,      document_range_formatting = false,      document_symbol = true,      find_references = true,      goto_definition = true,      hover = true,      implementation = true,      signature_help = false,      signature_help_trigger_characters = {},      text_document_did_change = 1,      text_document_open_close = true,      text_document_save = false,      text_document_save_include_text = false,      text_document_will_save = false,      text_document_will_save_wait_until = false,      workspace_symbol = false    },    rpc = {      handle = <userdata 1>,      notify = <function 13>,      pid = 8745,      request = <function 14>    },    server_capabilities = {      alsCalledByProvider = true,      alsReferenceKinds = { "reference", "write", "call", "dispatching call", "parent", "child" },      codeActionProvider = vim.empty_dict(),      completionProvider = {        resolveProvider = false,        triggerCharacters = { "." }      },      declarationProvider = true,      definitionProvider = true,      documentSymbolProvider = true,      executeCommandProvider = {        commands = { "als-named-parameters" }      },      foldingRangeProvider = true,      hoverProvider = true,      implementationProvider = true,      referencesProvider = true,      renameProvider = vim.empty_dict(),      textDocumentSync = 1,      typeDefinitionProvider = true    },    stop = <function 15>,    workspace_did_change_configuration = <function 16>  } }
[ DEBUG ] 2020-03-17T06:31:06Z+0100 ] ...nt_nvim7cMn6y/usr/share/nvim/runtime/lua/vim/lsp/rpc.lua:258 ]	"rpc.send.payload"	{  id = 2,  jsonrpc = "2.0",  method = "shutdown"}
[ DEBUG ] 2020-03-17T06:31:06Z+0100 ] ....mount_nvim7cMn6y/usr/share/nvim/runtime/lua/vim/lsp.lua:57 ]	"wait.fallback"	500
[ DEBUG ] 2020-03-17T06:31:06Z+0100 ] ...nt_nvim7cMn6y/usr/share/nvim/runtime/lua/vim/lsp/rpc.lua:338 ]	"decoded"	{  id = 2,  jsonrpc = "2.0",  result = vim.NIL}
[ DEBUG ] 2020-03-17T06:31:06Z+0100 ] ...nt_nvim7cMn6y/usr/share/nvim/runtime/lua/vim/lsp/rpc.lua:269 ]	"rpc.notify"	"exit"	nil
[ DEBUG ] 2020-03-17T06:31:06Z+0100 ] ...nt_nvim7cMn6y/usr/share/nvim/runtime/lua/vim/lsp/rpc.lua:258 ]	"rpc.send.payload"	{  jsonrpc = "2.0",  method = "exit"}
```

File: lua/nvim_lsp/als.lua [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  end
+  return X
+end
+
+local installer = make_installer()
+
+configs[server_name] = {
+  default_config = {
+    cmd = {bin_name};
+    filetype = {"ada"};
```
Comment (h-michael) [inline lua/nvim_lsp/als.lua]:
```suggestion
    filetypes = {"ada"};
```

Comment (h-michael) [inline lua/nvim_lsp/als.lua]:
```suggestion
    filetypes = {"ada"};
```
ref: #174

**Comment by h-michael** [comment]
@glacambre 
> Attempting to write a third procedure a3 below a2 results in the following error messages being printed on each keypress:

I can't reproduce it. Do you still face it?

**Comment by glacambre** [comment]
Yes, both with the most recent neovim appimage and building from source. It looks like `diagnostic.severity` is nil in `buf_diagnostics_underline` in `runtime/lua/vim/lsp/util.lua`. I'll keep investigating.

Do you have tips/tricks to debug lua in neovim?

**Comment by glacambre** [comment]
@h-michael I've added support for Windows and OSX (and was able to test on Windows but not on mac). I've also updated the Readme in order to mention the ALS. Please let me know if there's anything else I should do :).

**Comment by h-michael** [comment]
@glacambre Thanks :)

**Comment by glacambre** [comment]
Thanks for merging this and for all the work you, tjdevries and norcalli did on Neovim's LSP! :)

**Pull request merged**